### ISSUE TITLE
AI Gate 1A: fix medel tools and correnction control contract

### ISSUE THREAD

**Issue by emotigom** [AI-HealthCare-05/AH_05_07#204]
SK7의 `입력 기반 위험군 선별 신호`는 검증된 공개 데이터, 동결된 데이터 분할, 재현 가능한 모델 비교, 불변 모델 아티팩트를 전제로 한다.

현재 저장소에는 NHANES 데이터 계약과 모델 관련 스크립트가 존재하지만, 실제 사용 도구와 설치 의존성이 일치하지 않는다.

* 현재 모델 파이프라인은 pandas, scikit-learn, joblib을 사용한다.
* Parquet 입출력에 필요한 엔진이 AI 직접 의존성으로 명시되어 있지 않다.
* 현재 코드에서 사용하지 않는 PyTorch·sentence-transformers 계열과 Redis 의존성이 남아 있다.
* 도구의 채택 이유, 정확한 버전 확인 위치, 대체 조건, 보류 도구가 하나의 문서로 고정되어 있지 않다.

도구가 조용히 교체되거나 업그레이드 과정에서 품질·재현성이 저하되지 않도록, 실제 모델 학습 전에 AI 도구 SSOT와 자동 검증 경계를 확정한다.

## 목표

1. SK7 모델 개발에 사용하는 공개 데이터, 오픈소스 패키지, 내부 스크립트와 검증 도구를 명시한다.
2. `pyproject.toml`과 `uv.lock`을 실제 파이프라인에 맞춘다.
3. 승인되지 않은 AI·큐·LLM 의존성이 다시 들어오면 CI에서 감지한다.
4. 향후 도구 교체는 별도 Issue, 근거, 동일 split 비교 결과 없이는 진행할 수 없게 한다.
5. 제출 문서와 후속 스케일업에서 같은 도구와 결정 근거를 다시 찾을 수 있게 한다.

## 채택 도구

| 구분         | 도구                                                        | 역할                               | 채택 이유                               |
| ---------- | --------------------------------------------------------- | -------------------------------- | ----------------------------------- |
| 공개 데이터     | CDC NHANES 2017–March 2020 Pre-pandemic                   | 모델 학습 원천 데이터                     | 공식 공개 데이터이며 변수와 조사 방법을 확인할 수 있음     |
| 실행 환경      | Python 3.13                                               | 데이터 처리·학습·검증·추론                  | 현재 저장소의 대상 런타임                      |
| 환경·잠금      | uv + `uv.lock`                                            | 의존성 해석과 정확한 버전 고정                | 새 환경에서 동일 패키지 조합을 재현                |
| 데이터 처리     | pandas                                                    | XPT 읽기, 병합, 결측 처리, 파생 테이블 생성     | 현재 내부 데이터 스크립트가 사용                  |
| Parquet 엔진 | pyarrow                                                   | train/validation/test 및 예측 결과 저장 | 분할 파일을 안정적으로 교환                     |
| 수치 연산      | NumPy·SciPy                                               | scikit-learn의 수치 계산 기반           | 직접 모델 로직을 새로 구현하지 않음                |
| 모델 개발      | scikit-learn                                              | 전처리·학습·비교·평가                     | 표형 데이터에 적합하고 CPU 재현성이 높음            |
| 기준 모델      | LogisticRegression                                        | 해석 가능한 기본 기준선                    | 복잡한 모델의 성능 향상을 검증할 비교점              |
| 비교 모델      | HistGradientBoostingClassifier                            | 비선형 후보 모델                        | 같은 scikit-learn 안에서 의존성 증가 없이 비교 가능 |
| 전처리        | Pipeline, ColumnTransformer, SimpleImputer, OneHotEncoder | 학습·추론 변환 일치                      | 전처리 누락과 순서 불일치 방지                   |
| 모델 평가      | AUROC, PR-AUC, Brier score, calibration, 성별·연령대 subgroup  | 성능·확률 품질·그룹별 결과 검토               | 정확도 하나만으로 모델을 선택하지 않기 위함            |
| 아티팩트       | joblib + SHA-256 metadata                                 | 모델 저장 및 무결성 검증                   | 현재 CPU 모델 실행 계약과 일치                 |
| 테스트·정적 검사  | pytest, coverage, Ruff, mypy                              | 코드·계약 검증                         | 기존 CI 체계 재사용                        |
| 변경 기록      | GitHub Issue·PR·squash commit                             | 결정과 실행 근거 보존                     | 버전 변경 및 도구 교체 이력 추적                 |
| 구현 보조      | OpenAI Codex Work                                         | 코드·문서·테스트 작성 보조                  | 런타임이나 모델 결과 생성 도구로 사용하지 않음          |

정확한 패키지 버전은 `uv.lock`을 실행 기준으로 삼는다. 문서에는 선언 범위와 잠금 파일 확인 방법을 함께 기록한다.

## 내부 도구

다음 저장소 도구를 공식 AI 파이프라인으로 등록한다.

* `data/manifest/nhanes_2017_2020.json`

  * 데이터셋, 원본 파일, 라벨, 허용 변수, 금지 변수, split seed 기록
* `scripts/data/verify_manifest.py`

  * manifest 계약 검사
* `scripts/data/audit_schema.py`

  * 로컬 NHANES 파일과 필수 변수 검사
* `scripts/data/build_derived_table.py`

  * BP 라벨 변수와 예측 변수를 분리한 학습 테이블 생성
* `scripts/data/freeze_split.py`

  * train/validation/test 분할과 split digest 생성
* `scripts/model/compare_baselines.py`

  * 로지스틱 회귀와 histogram gradient boosting 비교
* `scripts/model/evaluate_predictions.py`

  * 전체·calibration·성별·연령대 결과 생성
* `scripts/model/train_artifact.py`

  * 승인 모델 아티팩트와 metadata 생성
* `app/core/model_artifact.py`

  * feature order, split digest, artifact hash 검증
* `app/core/model_registry.py`

  * 검증된 아티팩트 준비 여부 확인
* `app/core/model_runner.py`

  * 고정된 feature contract를 사용한 CPU 추론
* `docs/data-contract.md`

  * 데이터와 누수 방지 계약
* `docs/model-promotion.md`

  * 후보 모델 승격 조건

## 의존성 정리

* AI 그룹에 `pandas`, `pyarrow`, `scikit-learn`, `joblib`을 직접 선언한다.
* 사용 근거가 없는 다음 의존성을 제거한다.

  * `torch`
  * `torchvision`
  * `torchaudio`
  * `sentence-transformers`
* 실제 코드 사용과 승인 ADR이 없는 `redis` 의존성을 제거한다.
* PyTorch 전용 패키지 인덱스 설정을 제거한다.
* `uv.lock`을 갱신한다.

## 도구 변경 통제

다음 변경은 별도 Issue와 PR 없이는 허용하지 않는다.

* 공개 데이터셋 교체
* 라벨 정의 변경
* predictor 추가·삭제
* split seed 또는 비율 변경
* 전처리 방식 변경
* 모델 계열 변경
* 평가 지표 변경
* 라이브러리 주요 버전 업그레이드
* 아티팩트 저장 형식 변경
* Redis, Worker, LLM, OCR 또는 외부 모델 서비스 추가

도구나 모델을 교체하려면 다음 근거를 함께 제출한다.

1. 교체 이유
2. 기존 도구의 한계
3. 동일한 frozen validation split 비교
4. AUROC·PR-AUC·Brier·calibration 변화
5. subgroup 결과 변화
6. 실행시간·메모리·배포 영향
7. 재현 명령
8. rollback 방법
9. 안전·개인정보·라이선스 영향

## 신규 문서와 검사

* `docs/ai-toolchain-ssot.md`

  * 현재 채택·보류·금지 도구
  * 각 도구의 역할과 채택 이유
  * 정확한 버전 확인 위치
  * 내부 스크립트 실행 순서
  * 업그레이드·교체 절차
  * 제출 문서에서 인용할 위치
* `docs/adr/0002-ai-toolchain-and-change-control.md`

  * 도구 선택과 변경 통제 결정
* `scripts/ci/verify_ai_toolchain.py`

  * 허용된 AI 직접 의존성 확인
  * 미승인 PyTorch·sentence-transformers·Redis 재유입 감지
  * 필수 내부 스크립트와 문서 존재 확인
  * self-test 제공
* GitHub Actions

  * 기존 CI에 AI toolchain verifier 추가
* `docs/project-handoff.md`

  * 후속 AI 작업의 시작 지점을 새 SSOT로 연결

## 보류 도구

다음 도구는 이번 단계에서 채택하지 않는다.

* XGBoost·LightGBM
* TensorFlow·PyTorch
* Optuna
* SHAP
* MLflow
* Evidently
* Prometheus·Grafana
* Redis·Celery·별도 AI Worker
* LLM·sentence-transformers·외부 생성형 AI API

향후 동일 split에서 검증된 필요가 생기면 별도 ADR과 Issue로 검토한다.

## 범위 밖

* NHANES 원본 데이터 다운로드 또는 Git 커밋
* 실제 모델 학습과 성능 수치 생성
* 모델 아티팩트 생성·업로드·배포
* 위험 신호 API 공개
* 사용자 입력 폼 또는 결과 화면 연결
* Supabase 스키마 변경
* Cloud Run·Cloudflare·R2 변경
* 실제 사용자·건강 데이터 사용
* 모니터링 페이지 구현

## 안전 경계

* 사용자 표현은 `입력 기반 위험군 선별 신호`로 고정한다.
* 진단·치료·예방·인과적 개선을 주장하지 않는다.
* 라벨을 만드는 혈압 측정값과 그 파생값은 predictor로 사용하지 않는다.
* 이름, 연락처, 자유서술 병력, 실제 임상기록, JWT, 요청 본문을 모델 입력·로그·Git에 넣지 않는다.
* NHANES 원본과 파생 데이터는 로컬 작업 영역에만 두고 Git에 커밋하지 않는다.
* Codex 또는 다른 생성형 도구가 생성한 성능 수치를 근거로 사용하지 않는다. 모든 성능 값은 고정된 실행 스크립트의 산출물이어야 한다.

## 인수 조건

* [ ] `docs/ai-toolchain-ssot.md`에서 공개 데이터·오픈소스·내부 도구를 모두 확인할 수 있다.
* [ ] 각 채택 도구에 역할, 이유, 버전 확인 위치, 대체 조건이 기록되어 있다.
* [ ] `pyproject.toml`의 AI 그룹이 실제 모델 코드와 일치한다.
* [ ] `uv.lock`이 갱신되고 frozen 설치가 가능하다.
* [ ] 미사용 PyTorch·sentence-transformers·Redis 의존성이 제거된다.
* [ ] AI toolchain verifier의 정상·실패 self-test가 통과한다.
* [ ] CI가 승인되지 않은 의존성 재유입을 거부한다.
* [ ] 기존 앱·API·웹 테스트와 빌드가 통과한다.
* [ ] 모델 아티팩트가 없는 상태에서는 계속 `503 model_not_ready`를 반환한다.
* [ ] 모델 결과·DB·운영 배포에는 변화가 없다.

## 검증

```bash
uv lock --check
uv sync --group app --group dev --group ai --frozen
uv run python scripts/ci/verify_ai_toolchain.py --self-test
uv run python scripts/ci/verify_ai_toolchain.py
uv run ruff check .
uv run ruff format . --check
uv run coverage run -m pytest app
uv run coverage report -m
cd web
npm ci
npm run build
cd ..
python3 scripts/ci/verify_secret_boundary.py --self-test
python3 scripts/ci/verify_secret_boundary.py --web-dist web/dist
git diff --check
```

## 후속 작업

이 Issue 병합 후 별도 Issue에서 다음 순서로 진행한다.

1. NHANES 원본 파일·변수·사용 조건 감사
2. 누수 방지 파생 테이블 생성
3. train/validation/test split 동결
4. 두 기준 모델 비교
5. calibration·subgroup 평가
6. 모델 카드와 승격 결정
7. 불변 아티팩트 생성
8. 반복 입력·API 통합 검증
9. 사용자 입력과 분리된 위험 신호 이력 화면 연결

Notion에는 이 Issue가 병합된 뒤 GitHub PR과 squash commit을 근거로 도구 SSOT와 후속 순서를 동기화한다.
